src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,295 @@
+            // and at least 2 if there is also an IFormatProvider.  The result can be zero if 
+            // calling string.Format("String only with no placeholders"), where there is an 
+            // empty params array.
+            var numberOfPlaceholderArguments = numberOfArguments - (hasIFormatProvider ? 2 : 1);
+
+            var formatString = GetLiteralExpressionSyntaxAsString(formatStringLiteralExpressionSyntax);
+
+            if (FormatCallWorksAtRuntime(formatString, numberOfPlaceholderArguments))
+            {
+                return;
+            }
+            
+            ValidateAndReportDiagnostic(context, numberOfPlaceholderArguments,
+                formatString, GetLiteralExpressionSyntaxSpanStart(formatStringLiteralExpressionSyntax));
+        }
+
+        private string GetMethodName(ISyntaxFactsService syntaxFacts, SyntaxNode expression)
+        {
+            if (syntaxFacts.IsSimpleMemberAccessExpression(expression))
+            {
+                var nameOfMemberAccessExpression = syntaxFacts.GetNameOfMemberAccessExpression(expression);
+                return syntaxFacts.GetIdentifierOfSimpleName(nameOfMemberAccessExpression).ValueText;
+            }
+
+            if (syntaxFacts.IsIdentifierName(expression))
+            {
+                return syntaxFacts.GetIdentifierOfSimpleName(expression).ValueText;
+            }
+
+            return null;
+        }
+
+        private bool ArgsIsArrayOfReferenceTypes(
+            SemanticModel semanticModel,
+            SeparatedSyntaxList<SyntaxNode> arguments,
+            ImmutableArray<IParameterSymbol> parameters)
+        {
+            if (!TryGetArgsArgumentType(semanticModel, arguments, parameters, out var argsArgumentType))
+            {
+                return false;
+            }
+
+            if (argsArgumentType.TypeKind != TypeKind.Array)
+            {
+                return false;
+            }
+
+            var argsArray = (IArrayTypeSymbol)argsArgumentType;
+            return argsArray.ElementType.IsReferenceType;
+        }
+
+        internal abstract bool TryGetArgsArgumentType(
+            SemanticModel semanticModel,
+            SeparatedSyntaxList<SyntaxNode> arguments,
+            ImmutableArray<IParameterSymbol> parameters,
+            out ITypeSymbol argsArgumentType);
+
+        protected static bool TryGetValidFormatMethodSymbol(
+            SyntaxNodeAnalysisContext context,
+            int numberOfArguments,
+            SymbolInfo symbolInfo,
+            out IMethodSymbol method)
+        {
+            method = null;
+
+            if (symbolInfo.Symbol == null)
+            {
+                return false;
+            }
+
+            if (symbolInfo.Symbol.Kind != SymbolKind.Method)
+            {
+                return false;
+            }
+
+            if (!(symbolInfo.Symbol.ContainingSymbol is INamedTypeSymbol containingType))
```

Comment:
I think we determined that as long as `symbolInfo.Symbol` is an `IMethodSymbol`, then `ContainingSymbol` will never be null.